### DISCUSSION TITLE
when compress a folder on linux,  a strange error occured.

### DISCUSSION THREAD

**Discussion by flutelifh1980** [rikyoz/bit7z#214] [Q&A] [answered]
an error appeared as below. the 7z.so i used the  /usr/lib/p7zip/7z.so. and when the code run on  windows, it works.

version: 
7zip: 23.01
bit7z: release/v4.0.0
Linux version 5.10.0-5-generic (buildd@localhost) (gcc (Ubuntu 9.3.0-10kylin2) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #15~v10pro-KYLINOS SMP Tue Aug 3 03:55:56 UTC 2021

the code is c++ on linux:
code is as below, when run to Line 57, it crashed.  The linux system sent a signal. I don't know the reason, because the code can work on windows. 
![code](https://github.com/rikyoz/bit7z/assets/47734067/4db7dd7a-120c-4abf-b563-ba66477f8b90)

**Comment by rikyoz** [comment, marked as answer]
Hi!
By default bit7z builds for 7-Zip 23.01. Unfortunately, this version of 7-Zip introduced some incompatibilities with p7zip's shared libraries that can cause problems like yours.
If you want to use p7zip's 7z.so with bit7z, you need to build the latter with the `BIT7Z_USE_LEGACY_IUNKNOWN` option enabled.
This should solve your problem.

**Reply by flutelifh1980** [reply]
ok. i build it latter and test again. thanks!

**Reply by flutelifh1980** [reply]
hi,
  i add code like this in CMakeList.txt of bit7z to build the libray on linux :  add_definitions("-DBIT7Z_USE_LEGACY_IUNKNOWN") 
 finally, i found  the error message appeared like before when my code runs. 
what's wrong with it ?  
 waiting your reply.
   thanks!

**Reply by rikyoz** [reply]
`add_definitions` can be used to define new C++ preprocessor definitions, not to enable CMake options like `BIT7Z_USE_LEGACY_IUNKNOWN`.

In this case, the way you enable it depends on how you're building bit7z.

If you're including bit7z in your project via `add_subdirectory`, you need to set it like this
```cmake
set(BIT7Z_USE_LEGACY_IUNKNOWN ON CACHE BOOL "use legacy iunknown" FORCE)
```
_before_ bit7z's `add_subdirectory` in your CMakeLists.txt.

If instead you're building bit7z separately from your project, you need to pass `-DBIT7Z_USE_LEGACY_IUNKNOWN=ON` to the CMake configuration command of bit7z.
